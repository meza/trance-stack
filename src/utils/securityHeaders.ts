import { contentSecurityPolicy } from '~/utils/contentSecurityPolicy';

// https://owasp.org/www-project-secure-headers/ci/headers_remove.json
const headersToClear = {
  'last_update_utc': '2023-02-05 19:51:17',
  'headers': [
    'Host-Header',
    'K-Proxy-Request',
    'Liferay-Portal',
    'Pega-Host',
    'Powered-By',
    'Product',
    'Server',
    'SourceMap',
    'X-AspNet-Version',
    'X-AspNetMvc-Version',
    'X-Atmosphere-error',
    'X-Atmosphere-first-request',
    'X-Atmosphere-tracking-id',
    'X-B3-ParentSpanId',
    'X-B3-Sampled',
    'X-B3-SpanId',
    'X-B3-TraceId',
    'X-CF-Powered-By',
    'X-CMS',
    'X-Content-Encoded-By',
    'X-Envoy-Attempt-Count',
    'X-Envoy-External-Address',
    'X-Envoy-Internal',
    'X-Envoy-Original-Dst-Host',
    'X-Envoy-Upstream-Service-Time',
    'X-Framework',
    'X-Generated-By',
    'X-Generator',
    'X-Mod-Pagespeed',
    'X-Page-Speed',
    'X-Php-Version',
    'X-Powered-By',
    'X-Powered-By-Plesk',
    'X-Powered-CMS',
    'X-Redirect-By',
    'X-Server-Powered-By',
    'X-SourceFiles',
    'X-SourceMap',
    'X-Turbo-Charged-By',
    'X-Varnish-Backend',
    'X-Varnish-Server'
  ]
};

export const securityHeaders = (isDev: boolean): Headers => {
  const owaspHeaders = new Headers();

  owaspHeaders.set('Content-Security-Policy', contentSecurityPolicy(isDev));
  owaspHeaders.set('Cross-Origin-Embedder-Policy', 'require-corp');
  owaspHeaders.set('Cross-Origin-Opener-Policy', 'same-origin');
  owaspHeaders.set('Cross-Origin-Resource-Policy', 'same-origin');
  owaspHeaders.set('Origin-Agent-Cluster', '?1');
  owaspHeaders.set('Permissions-Policy', 'accelerometer=(), camera=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), payment=(), usb=()');
  owaspHeaders.set('Referrer-Policy', 'origin-when-cross-origin');
  owaspHeaders.set('Strict-Transport-Security', 'max-age=15552000; includeSubDomains');
  owaspHeaders.set('X-Content-Type-Options', 'nosniff');
  owaspHeaders.set('X-DNS-Prefetch-Control', 'off');
  owaspHeaders.set('X-Download-Options', 'noopen');
  owaspHeaders.set('X-Frame-Options', 'SAMEORIGIN');
  owaspHeaders.set('X-Permitted-Cross-Domain-Policies', 'none');
  owaspHeaders.set('X-XSS-Protection', '0');

  return owaspHeaders;
};

export const addSecurityHeaders = (headers: Headers, isDev: boolean): Headers => {
  const owaspHeaders = securityHeaders(isDev);

  owaspHeaders.forEach((value, key) => {
    headers.set(key, value);
  });

  return headers;
};

export const sanitizeHeaders = (headers: Headers): Headers => {
  headersToClear.headers.forEach((header) => {
    headers.delete(header);
  });

  return headers;
};
