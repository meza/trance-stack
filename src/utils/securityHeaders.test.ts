import { beforeEach, describe, expect, it, vi } from 'vitest';
import { addSecurityHeaders, sanitizeHeaders, securityHeaders } from '~/utils/securityHeaders';

describe('securityHeaders', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('sets the correct security headers', () => {
    const existingHeaders = new Headers();
    const headers = securityHeaders('noncevalue');
    const modifiedHeaders = addSecurityHeaders(existingHeaders, 'noncevalue');

    expect(headers).toStrictEqual(modifiedHeaders);
    expect(headers).toStrictEqual(existingHeaders);

    expect(Object.fromEntries(headers.entries())).toMatchSnapshot();
  });

  it('removes unwanted headers', () => {
    const existingHeaders = new Headers();
    existingHeaders.set('x-powered-by', 'Vite');
    existingHeaders.set('x-vite', 'true');

    const sanitized = sanitizeHeaders(existingHeaders);

    expect(sanitized).toBe(existingHeaders);
    expect(existingHeaders.has('x-powered-by')).toBeFalsy();
    expect(existingHeaders.has('x-vite')).toBeTruthy();
  });
});
