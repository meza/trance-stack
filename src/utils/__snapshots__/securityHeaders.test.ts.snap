// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`securityHeaders > sets the correct security headers 1`] = `
{
  "content-security-policy": "script-src 'nonce-noncevalue' 'strict-dynamic'; object-src 'none'; base-uri 'none';",
  "cross-origin-embedder-policy": "same-origin",
  "cross-origin-opener-policy": "same-origin",
  "cross-origin-resource-policy": "same-origin",
  "origin-agent-cluster": "?1",
  "permissions-policy": "accelerometer=(), camera=(), geolocation=(), gyroscope=(), magnetometer=(), microphone=(), payment=(), usb=()",
  "referrer-policy": "origin-when-cross-origin",
  "strict-transport-security": "max-age=31536000; includeSubDomains",
  "x-content-type-options": "nosniff",
  "x-dns-prefetch-control": "off",
  "x-download-options": "noopen",
  "x-frame-options": "SAMEORIGIN",
  "x-permitted-cross-domain-policies": "none",
  "x-xss-protection": "0",
}
`;
